### DISCUSSION TITLE
Leaving the Connect dialog always open, what are your thoughts ?

### DISCUSSION THREAD

**Discussion by dcorson-ticino-com** [jamulussoftware/jamulus#1176] [UI Related]
Leaving the Connect dialog always open, what are your thoughts ?
Is the network usage significant enough to restrict this?
I see added network usage for the client of about 24kbs long term with short term peaks of over 100kbs.
On my server I see a difference of under 2kbs adding one client.
As I see it the client will never be pinging more than 150 servers, but still.
The server can be faced with clients without limit, however (or is there a limit?).
And where is the directory server in this game?

I have seen several people advocating the Connect menu always open, and it is possible now, probably not used by most.  Just wondering as I am now considering an UI where all windows are always open.
(A possibility to reduce traffic would be to reduce the update frequency when the mouse is not over the open window.  The opposite of now.)

**Comment by gene96817** [comment]
As our local evangelist, I like having the Connect dialog always open. It gives me a view of the activity on our servers. I am interested in seeing when there are visitors to our servers. I'd like to believe that +/- 100 kbps of traffic is not significant (and yes, some people are operating at the margins).

**Comment by ann0see** [comment]
As far as I know having it open causes drop outs (at least that's what corrados wrote somewhere).

**Comment by hoffie** [comment]
Hm, personally I would not need this as I'm mostly connected to a single server. I understand that people who use Jamulus for actual jamming might want to have a permanent view on the activity on other servers or that local power users want to monitor neighbor servers.

In my opinion, we should not show this by default. We could add a "keep connect dialog open" option if we really wanted to. Besides that, simply running second (non-connected) Jamulus instance would solve this as well, wouldn't it?

**Reply by gene96817** [reply]
We don't need "keep connect dialog open" to be a default. Better to close it for most users so that it is not a distraction.  

> Besides that, simply running a second (non-connected) Jamulus instance would solve this as well, wouldn't it?

Would that take more resources than just a single Jamulus instance? On one hand, we were trying to save small increases of resource usage. Just starting up another instance feels like, hey resources are free.

**Comment by gilgongo** [comment]
My understanding is that the fact that the window polls the directory means that the directory server may have problems if lots of clients had their windows open. Also in some cases it may degrade client performance if the user's router can't handle (fragmenting?) UDP packets fast enough. 

I've advocated auto-refresh once on opening the window, and having a "refresh" button on the window so that it can be left up.

**Reply by gene96817** [reply]
Does all connection window refresh come from the directory? I see some comments that imply the client also polls the servers in their local list (which originally comes from the directory). This makes more sense to me. We could have two refresh cycles. (1) the refresh going directly to the servers and (2) a much slower refresh to the directory. The only reason to get an update from the directory is to discover a new server that just comes online. All the other information would be discovered from the direct updates.

**Reply by dcorson-ticino-com** [reply]
Of course the directory can not know the ping times from a certain client to the servers.
First the server list is transferred from the directory server to the client and then the client queries the individual servers.
I am going to look at the logic in detail in the coming days.

**Comment by dcorson-ticino-com** [comment]
The client Connect dialog works like this:
1) The directory server is queried for its server list.  This query is repeated every 2 sec until a response is seen.  After that there is no more contact with the directory server.
2) Pings are sent to all the servers on the list every 2 seconds.  The list is updated as the responses arrive.  This continues until the window is closed.

This means, for example, that if you open the Connect window and let it run all day you will never see any new servers as the directory is only queried once.
I will now implement a Connect window with Update button.  The behaviour will be:
1. Query directory as above.
2. Send pings to the listed servers every 2 sec for 8 times.  It appears to me that the list settles down at around 6 ping requests.  After 8 requests the pings stop, no further communication.
3. Clicking Update starts the procedure at 1) as does opening the Connect window in any manner such as disconnecting.
Updating is not necessary to select and connect to a server in the list.
Double clicking on a server while connected closes the connection and opens the new one.

We will see how that looks and handles,
Don

**Reply by gene96817** [reply]
Thank you for the description.

I suggest a slow ping rate to the directory (5 minutes) to get new servers coming online. Possibly just a manual refresh is just as effective.

A ping refresh of the Connect window would be helpful. every 60 seconds. I'd like to have this to monitor parallel activity (e.g. sectionals) on other servers.
